google-cloud-spanner/src/main/java/com/google/cloud/spanner/XGoogSpannerRequestId.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2025 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.spanner;
+
+import com.google.api.core.InternalApi;
+import com.google.common.annotations.VisibleForTesting;
+import java.security.SecureRandom;
+import java.util.Objects;
+
+@InternalApi
+public class XGoogSpannerRequestId {
+  // 1. Generate the random process Id singleton.
+  @VisibleForTesting static String RAND_PROCESS_ID = XGoogSpannerRequestId.generateRandProcessId();
+
+  @VisibleForTesting
+  static final long VERSION = 1; // The version of the specification being implemented.
+
+  private final long nthClientId;
+  private final long nthChannelId;
+  private final long nthRequest;
+  private long attempt;
+
+  XGoogSpannerRequestId(long nthClientId, long nthChannelId, long nthRequest, long attempt) {
+    this.nthClientId = nthClientId;
+    this.nthChannelId = nthChannelId;
+    this.nthRequest = nthRequest;
+    this.attempt = attempt;
+  }
+
+  public static XGoogSpannerRequestId of(
+      long nthClientId, long nthChannelId, long nthRequest, long attempt) {
+    return new XGoogSpannerRequestId(nthClientId, nthChannelId, nthRequest, attempt);
+  }
+
+  private static String generateRandProcessId() {
+    // Expecting to use 64-bits of randomness to avoid clashes.
+    return String.format("%016x", new SecureRandom().nextInt(64));
```

Comment:
The `bound` argument to `nextInt(int)` does not indicate the number of bits to use, it indicates the highest number that should be generated. Meaning: This call will return a random value in the range `[0, 64)`. Also, an `int` in Java is by definition 32-bit, so the length of the hexadecimal representation of it will never be more than 8 characters (and the spec for the random process ID is that it should be a 32-bit unsigned int). So this should just be:

```suggestion
    return String.format("%08x", new SecureRandom().nextInt());
```

This will:
1. Generate a random 32-bit integer from the full 32-bit range (so including negative numbers)
2. The `%08x` format will convert it to the corresponding __unsigned__ hex format (so `-1` is formatted as `ffffffff`)

Reply:
Hey @olavloite thanks for the discourse and for pointing that out. I instead went with the use of BigInteger plus SecureRandom to generate a 64-bit integer. Also as per https://orijtech.notion.site/x-goog-spanner-request-id-always-on-gRPC-header-to-aid-in-quick-debugging-of-errors-14aba6bc91348091a58fca7a505c9827#14aba6bc9134807f9f17d3ef58e26128 I discussed the problems of using a 32-bit random integer and potential clashes, meanwhile 64-bit random integer is the right balance and also in our Google group.